### ISSUE TITLE
problem with input

### ISSUE THREAD

**Issue by DZ4VR** [qubvel/segmentation_models#32]
Hello! I have some problems with using unet tamplate
If I try to use unet with
x, y = X_train,y_train
(X_train contains np.array of images as well as y_train and their shapes are (10,512, 512, 3) - 10 is count of samples 512x512 is resolution and 3 is rgb)
But I get  this error :Error when checking target: expected sigmoid to have shape (None, None, 1) but got array with shape (512, 512, 3)

**Comment by chm60** [comment]
Have you specified the input shape when you define the Model.

`
model = Unet(backbone_name='resnet18', encoder_weights='imagenet', classes=59,
        input_shape=(512,512,3),  activation='softmax')
`

**Comment by DZ4VR** [comment]
@chm60 Ok i just changed y from (10,512,512,3) to (10,512,512,1) but now I have problem with prediction: output shape is (10,512,512,1) but I wanted just (512,512,1). In my X and y 10 is just amount of samples, how can I fix it?

**Comment by chm60** [comment]
can you post your code please :)

**Comment by DZ4VR** [comment]
`import glob
filenames = glob.glob('aug/*')
filenames = sorted(filenames)
pics = []
masks = []
for name in filenames:
    if name[-8:-4] == 'mask':
        masks.append(name)
    else:
        pics.append(name)
df_aug = pd.DataFrame({'img':pics,'mask':masks})
for i in range(169):
    df_aug['img'][i] = cv2.imread (df_aug['img'][i], 3)
    df_aug['mask'][i] = cv2.imread (df_aug['mask'][i], 1)
for i in range(169):
    df_aug['mask'][i]=df_aug['mask'][i][..., :1]
X_train = np.array(np.array(df_aug['img']).tolist())
y_train = np.array(np.array(df_aug['mask']).tolist())
X_train10 = X_train[:10][:][:][:]
y_train10 = y_train[:10][:][:][:]
from segmentation_models import Unet
from segmentation_models.backbones import get_preprocessing

# prepare data
x, y = X_train10,y_train10

preprocessing_fn = get_preprocessing('resnet34')
x = preprocessing_fn(x)

# prepare model
model = Unet(backbone_name='resnet34', encoder_weights='imagenet')
model.compile('Adam', 'binary_crossentropy', ['binary_accuracy'])

# train model
print(x.shape)
model.fit(x, y)`

**Comment by chm60** [comment]
I think you're getting confused with what the .fit function actually does.
So in the code above you're training the model with 10 samples of data and iterating it once ( epoch ).

to return a prediction simply put :
`result = model.predict(data)`


where data is another image, as an array of size 512,512,3. Also i'd advice to configure your model to below:


`NUMBER_OF_CLASSES = 1
model = Unet(backbone_name='resnet34', encoder_weights='imagenet', input_shape=(512,512,3), classes=NUMBER_OF_CLASSES)
`

**input shape** is the shape of your images without the amount of samples you're passing through, and the **classes** keyword refers to how many different objects which you want the model to recognise could potentially be in the image.

**Comment by DZ4VR** [comment]
@chm60 Thank you! And by the way dont I need to implement batches? Can I just throw all my data in x,y?

**Comment by chm60** [comment]
There's a method in model.fit which takes care of batch size for you.

`model.fit(x,y, batch_size=10)`

If you intend to train the model with a lot of data, for example i trained it with 1004 images (1004,512,512,3)  and subsequently their masks which were (1004, 512, 512, 59) I had memory issues.
For future reference you should look into generators to load data from disk to memory in batches and then into the model.

This link provides more details about the methods available for Keras models
https://keras.io/models/model/

This link provides information regarding generators to prevent memory issues in the future ( if you have a large data set )
https://medium.com/@ensembledme/writing-custom-keras-generators-fe815d992c5a

Keep at it and keep learning :smile:

**Comment by DZ4VR** [comment]
@chm60 And another question: does it understands that shape[0] is just amount of pics?
